libnopegl/src/text.c

```diff
@@ -209,17 +435,131 @@ int ngli_text_set_string(struct text *s, const char *str)
         }
     }
 
+    /* Default data for each character */
+    const struct {
+        float transform[4 * 4];
+        float color[3];
+        float opacity;
+    } default_data = {
+        .transform = NGLI_MAT4_IDENTITY,
+        .color     = {NGLI_ARG_VEC3(s->config.defaults.color)},
+        .opacity   = s->config.defaults.opacity,
+    };
+
+    /*
+     * Reallocate characters data if the number of characters changed. We could
+     * use a "<" instead of "!=" but we don't want to keep large amount of
+     * memory allocated in case one live change event set a large string which
+     * is later trimed down.
+     */
```

Comment:
Can  we use a mechanism that allocates every time to the next power of 2 if needed_size > allocated_size ? I'm thinking about a scenario where the user is live editing a node and typing text on the flight, where a realloc would be triggered each type the number of chars increases. Unless i've missed something, the overall memory footprint is rather low compared to what we need when compositing video streams).

Reply:
Morever, it happens with the tools ui when the frame number char number increases or decreases, especially when the player seeks back from `duration` to `0`.

Reply:
Ah that's not a bad idea to stitch the allocation to a known power of two. I adjusted the code and extended the comment. (Note: I didn't bother using clz for that)

One thing though: I added a `chars_copy_size`, because we don't need to copy the next-power-of-two amount of data when we have less characters. Making this change made me realize a few short comings in the code, so I made a few other adjustments in that area:

1. `ngli_text_update_effects_defaults()` was not updating the buffers: this was revealed when the re-allocation didn't happen as often as before. It's now filling the default data buffers
2. `chars_data_default` now uses realloc instead of relying on `destroy_effects_data` + `calloc`
3. the default data buffers are now updated only when the number of character changes

We could be smarter to reduce a few more copies but I don't think it's worth the effort for now.

Reply:
Looks good ! Thanks.